src/main/webapp/app/entities/result/result.component.ts

```diff
@@ -51,54 +71,108 @@ export class ResultComponent implements OnInit, OnChanges {
-                        }
-                        if (r1.completionDate! < r2.completionDate!) {
-                            return 1;
-                        }
-                        return 0;
-                    });
-                }
+            if (this.hasParticipationResults()) {
+                const result = this.getLatestResult(this.participation.results);
+
                 // Make sure result and participation are connected
-                this.result = this.participation.results[0];
+                this.result = result;
                 this.result.participation = this.participation;
             }
 
-            this.init();
+            this.evaluate();
         }
     }
 
-    init() {
-        if (this.result && (this.result.score || this.result.score === 0) && (this.result.rated === true || this.result.rated == null || this.showUngradedResults)) {
+    evaluate() {
+        this.evaluateTemplateStatus();
+
+        if (this.result && (this.result.score || this.result.score === 0) && (this.result.rated || this.result.rated == null || this.showUngradedResults)) {
             this.textColorClass = this.getTextColorClass();
             this.hasFeedback = this.getHasFeedback();
             this.resultIconClass = this.getResultIconClass();
             this.resultString = this.buildResultString();
             this.resultTooltip = this.buildResultTooltip();
+        } else if (this.templateStatus === ResultTemplateStatus.LATE) {
+            this.textColorClass = 'result-gray';
+            this.resultIconClass = this.getResultIconClass();
         } else {
             // make sure that we do not display results that are 'rated=false' or that do not have a score
             this.result = null;
         }
     }
 
     ngOnChanges(changes: SimpleChanges) {
-        if (changes.participation || changes.result) {
-            this.ngOnInit();
+        if (changes.participation || changes.result || changes.participation) {
+            this.init();
+        }
+    }
+
+    evaluateTemplateStatus() {
+        if (this.isModelingOrTextOrFileUpload() && this.participation && this.participation.exercise) {
+            const assessmentDueDate = this.dateAsMoment(this.participation.exercise!.assessmentDueDate!);
+            if (this.isSubmissionInDueTime()) {
+                if (this.hasResultAndScore()) {
+                    // Prevent that result is shown before assessment due date
+                    if (!assessmentDueDate || assessmentDueDate.isBefore()) {
+                        this.templateStatus = ResultTemplateStatus.HAS_RESULT;
+                    } else {
+                        this.templateStatus = ResultTemplateStatus.NO_RESULT;
+                    }
+                } else {
+                    this.templateStatus = ResultTemplateStatus.SUBMITTED;
+                }
+            } else if (this.hasResultAndScore() && (!assessmentDueDate || assessmentDueDate.isBefore())) {
```

Comment:
I'm not a big fan of methods as `this.hasResultAndScore()`:
- `this` is the Angular component. Does it make sense that the angular component does not have a result and score?
- Has no arguments and is stateful (=untestable)
- It doesn't tell that not having a score includes `score === 0`

I would propose creating a util for this, we e.g. have `participation-utils` for similar cases.

Reply:
Created a util for it :-)